wallets/server.js

```diff
@@ -110,18 +110,47 @@ export async function createWrappedInvoice (userId,
   }
 }
 
-export async function getInvoiceableWallets (userId, { models }) {
-  const wallets = await models.wallet.findMany({
-    where: { userId, enabled: true },
-    include: {
-      user: true
-    },
-    orderBy: [
-      { priority: 'asc' },
-      // use id as tie breaker (older wallet first)
-      { id: 'asc' }
-    ]
-  })
+export async function getInvoiceableWallets (userId, { predecessorId, models }) {
+  // filter out all wallets that have already been tried by recursively following the retry chain of predecessor invoices.
+  // the current predecessor invoice is in state 'FAILED' and not in state 'RETRYING' because we are currently retrying it
+  // so it has not been updated yet.
+  // if predecessorId is not provided, the subquery will be empty and thus no wallets are filtered out.
+  const wallets = await models.$queryRaw`
+    SELECT
+      "Wallet".*,
+      jsonb_build_object(
+        'id', "users"."id",
+        'hideInvoiceDesc', "users"."hideInvoiceDesc"
+      ) AS "user"
+    FROM "Wallet"
+    JOIN "users" ON "users"."id" = "Wallet"."userId"
+    WHERE
+      "Wallet"."userId" = ${userId}
+      AND "Wallet"."enabled" = true
+      AND "Wallet"."id" NOT IN (
+        WITH RECURSIVE "Retries" AS (
+          -- select the current failed invoice that we are currently retrying
+          -- this failed invoice will be used to start the recursion
+          SELECT "Invoice"."id", "Invoice"."predecessorId"
+          FROM "Invoice"
+          WHERE "Invoice"."id" = ${predecessorId} AND "Invoice"."actionState" = 'FAILED'
+
+          UNION ALL
+
+          -- recursive part: use predecessorId to select the previous invoice that failed in the chain
+          -- until there is no more previous invoice
+          SELECT "Invoice"."id", "Invoice"."predecessorId"
+          FROM "Invoice"
+          JOIN "Retries" ON "Invoice"."id" = "Retries"."predecessorId"
+          WHERE "Invoice"."actionState" = 'RETRYING'
+        )
+        SELECT
+          "InvoiceForward"."walletId"
+        FROM "Retries"
+        JOIN "InvoiceForward" ON "InvoiceForward"."invoiceId" = "Retries"."id"
+        WHERE "InvoiceForward"."withdrawlId" IS NOT NULL
```

Comment:
Mhhh, I think `WHERE "Retries"."isHeld" = 't'` instead of `"InvoiceForward"."withdrawlId" IS NOT NULL` would also work and would not require a join :thinking:

Reply:
we can check `actionError` to be sure

Reply:
It is currently not set for forward failures but I can change that if you meant that

Reply:
I have a fix I'm working on locally. I just need to make sure I didn't regress it.

Reply:
I don't know if this changed or was always there, but fees are being deducted from p2p zap receives and shouldn't be

<img width="591" alt="Screenshot 2024-12-10 at 1 45 58 PM" src="https://github.com/user-attachments/assets/4b15a362-3937-4ace-956a-405fc70d9dae">

Reply:
Then my notification adds the fee too

<img width="552" alt="Screenshot 2024-12-10 at 1 47 49 PM" src="https://github.com/user-attachments/assets/3a190b00-4456-4f10-b611-cc5a254c5ea2">

Reply:
> but fees are being deducted from p2p zap receives and shouldn't be

Mhh, I thought I fixed that at some point. If the wallet created a 70 sat invoice, it should obviously receive 70 sats since senders pay network fees.

> Then my notification adds the fee too

This is also wrong, right? Imo, the notification should show how much you received and if you click on it, the fees are shown separately.

So for a 100 sat zap, it should show "70 sats were zapped directly to your attached wallet" and if you click into it, it will show "70 sats with 1 sat routing fee".

Does this make sense / do you agree?

Reply:
SN pays the fee from SN to the receiver. So it shouldn't show any fee.